### PR TITLE
tests: add blackbox test

### PR DESCRIPTION
**Pull request by arithx**
Adds a blackbox test to tests/blackbox_test. Also disables the
tests/ folder temporarily from being ran by the test script.

### CODE REVIEW

**Comment by dgonyeo** [comment]
This is a continuation of https://github.com/coreos/ignition/pull/375

File: tests/blackbox_test.go [hunk 1]
```diff
@@ -0,0 +1,1215 @@
...
+	size = size + int64(4096*512) // extra room to allow for alignments
+	return size
+}
+
+func createVolume(
+	t *testing.T, fileName string, size int64, cylinders int, heads int,
+	sectorsPerTrack int, partitions []*Partition) {
+	// attempt to create the file, will leave already existing files alone.
+	// os.Truncate requires the file to already exist
+	out, err := os.Create(fileName)
```
Comment (dgonyeo) [inline tests/blackbox_test.go]: I think you can combine the `os.Create` here and the `os.Truncate` below into one line with: `os.OpenFile(fileName, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)`

File: tests/blackbox_test.go [hunk 2]
```diff
@@ -0,0 +1,1215 @@
...
+	}
+
+	createPartitionTable(t, fileName, partitions)
+
+	for counter, partition := range partitions {
+		if partition.TypeCode == "blank" || partition.FilesystemType == "" {
+			continue
+		}
+
+		mntPath, err := ioutil.TempDir("", fmt.Sprintf("%s%d", "hd1p", counter))
```
Comment (dgonyeo) [inline tests/blackbox_test.go]: Tiny nit, why not `fmt.Sprintf("hd1p%d", counter)` ?

File: tests/blackbox_test.go [hunk 3]
```diff
@@ -0,0 +1,1215 @@
...
+}
+
+func formatPartition(t *testing.T, partition *Partition) {
+	switch partition.FilesystemType {
+	case "vfat":
+		formatVFAT(t, partition)
+	case "ext2", "ext4":
+		formatEXT(t, partition)
+	case "btrfs":
+		formatBTRFS(t, partition)
```
Comment (dgonyeo) [inline tests/blackbox_test.go]: Are the tests not covering XFS or swap partitions?

File: tests/blackbox_test.go [hunk 4]
```diff
@@ -0,0 +1,1215 @@
...
+func align(count int, alignment int) int {
+	offset := count % alignment
+	if offset != 0 {
+		count += alignment - offset
+	}
+	return count
+}
+
+func setOffsets(partitions []*Partition) {
+	offset := 34
```
Comment (dgonyeo) [inline tests/blackbox_test.go]: This number also feels somewhat magical. Can you comment it?

File: tests/blackbox_test.go [hunk 5]
```diff
@@ -0,0 +1,1215 @@
...
+
+func kpartxAdd(t *testing.T, fileName string) string {
+	kpartxOut, err := exec.Command(
+		"/sbin/kpartx", "-av", fileName).CombinedOutput()
+	if err != nil {
+		t.Fatal("kpartx", err, string(kpartxOut))
+	}
+	kpartxOut, err = exec.Command(
+		"/sbin/kpartx", "-l", fileName).CombinedOutput()
+	return strings.Trim(strings.Split(string(kpartxOut), " ")[4], "/dev/")
```
Comment (dgonyeo) [inline tests/blackbox_test.go]: This will panic if `kpartx` doesn't return the output you're expecting. Since this is in a test, I'm not sure this is a problem/worth fixing though.

File: tests/blackbox_test.go [hunk 6]
```diff
@@ -0,0 +1,1215 @@
...
+}
+
+func mountRootPartition(t *testing.T, partitions []*Partition) {
+	for _, partition := range partitions {
+		if partition.Label != "ROOT" {
+			continue
+		}
+		mountOut, err := exec.Command(
+			"/bin/mount", partition.Device,
```
Comment (dgonyeo) [inline tests/blackbox_test.go]: Tiny nit, feel free to ignore, for things like this I generally prefer to use [the `syscall` package](https://golang.org/pkg/syscall/#Mount) instead of exec'ing something else. Not really an issue here because this entire thing is already dependent on a bunch of other binaries.

File: tests/blackbox_test.go [hunk 7]
```diff
@@ -0,0 +1,1215 @@
...
+	for _, str := range strings {
+		if str != "" {
+			r = append(r, str)
+		}
+	}
+	return r
+}
+
+func generateUUID(t *testing.T) string {
+	out, err := exec.Command("/usr/bin/uuidgen").CombinedOutput()
```
Comment (dgonyeo) [inline tests/blackbox_test.go]: Any reason you're hardcoding the path of all the binaries in this? Why not let `exec` find it in the `$PATH`? On some OSs (*cough* NixOS *cough*) these won't be where you're expecting.

File: tests/blackbox_test.go (hunk 1 above)
Comment (dgonyeo) [inline tests/blackbox_test.go]: Oh wait, I now see that you're truncating it to a non-zero size, and after reading the docs for `os.Create` it says:

> Create creates the named file with mode 0666 (before umask), truncating it if it already exists

I think you'll want to call `OpenFile` with just `os.O_RDWR` and `os.O_CREATE` to prevent it from truncating it to 0, unless that isn't an issue here.

File: tests/blackbox_test.go [hunk 8]
```diff
@@ -0,0 +1,1215 @@
...
+			"/sbin/sgdisk", "-i", strconv.Itoa(e.Number),
+			fileName).CombinedOutput()
+		if err != nil {
+			t.Fatal("sgdisk -i", strconv.Itoa(e.Number), err)
+		}
+		lines := strings.Split(string(sgdiskInfo), "\n")
+		actualTypeGUID := strings.ToUpper(strings.TrimSpace(
+			strings.Split(strings.Split(lines[0], ": ")[1], " ")[0]))
+		actualSectors := strings.Split(strings.Split(lines[4], ": ")[1], " ")[0]
+		actualLabel := strings.Split(strings.Split(lines[6], ": ")[1], "'")[1]
```
Comment (dgonyeo) [inline tests/blackbox_test.go]: Same comment as earlier that this will panic if the output isn't what you expect, and that it might be fine.

File: tests/blackbox_test.go [hunk 9]
```diff
@@ -0,0 +1,1215 @@
...
+
+func validateFiles(t *testing.T, expected []*Partition) {
+	for _, partition := range expected {
+		if partition.Files == nil {
+			continue
+		}
+		for _, file := range partition.Files {
+			path := filepath.Join(partition.MountPath, file.Path, file.Name)
+			if _, err := os.Stat(path); os.IsNotExist(err) {
+				t.Fatal("File doesn't exist!", path)
```
Comment (dgonyeo) [inline tests/blackbox_test.go]: If you call `t.Error` instead of `t.Fatal` in places like this, when something goes wrong the tester will be able to see all the errors instead of just the first one that was found.

File: tests/blackbox_test.go [hunk 10]
```diff
@@ -0,0 +1,1215 @@
...
+				filepath.Join(p.MountPath, "usr", "lib64"), filepath.Join(
+					p.MountPath, "lib64"))
+			_ = os.Symlink(
+				filepath.Join(p.MountPath, "usr", "bin"), filepath.Join(
+					p.MountPath, "bin"))
+			_ = os.Symlink(
+				filepath.Join(p.MountPath, "usr", "sbin"), filepath.Join(
+					p.MountPath, "sbin"))
+
+			_, _ = exec.Command(
```
Comment (arithx) [inline tests/blackbox_test.go]: All of the libraries being pulled in are for Ignition itself. If they aren't present the `getpwnam_r` in `internal/exec/util/user_group_lookup.c` on line 61 will set `pptr` to null.

File: tests/blackbox_test.go (hunk 4 above)
Comment (arithx) [inline tests/blackbox_test.go]: Adding a comment to the code but 34 is first non-reserved sector for GPT

File: tests/blackbox_test.go (hunk 1 above)
Comment (arithx) [inline tests/blackbox_test.go]: If it were to already exist truncating it wouldn't be an issue. Going to be changing it to be a tmp file so it shouldn't run into that case anyways.

File: tests/blackbox_test.go (hunk 5 above)
Comment (arithx) [inline tests/blackbox_test.go]: Adding a nil check on the err with a Fatal call so it will properly fail out rather than panic.

File: tests/blackbox_test.go [hunk 11]
```diff
@@ -0,0 +1,1215 @@
...
+		}, {
+			Number:   2,
+			Label:    "BIOS-BOOT",
+			TypeCode: "bios",
+			Length:   4096,
+		}, {
+			Number:         3,
+			Label:          "USR-A",
+			GUID:           "7130c94a-213a-4e5a-8e26-6cce9662f132",
+			TypeCode:       "coreos-rootfs",
```
Comment (arithx) [inline tests/blackbox_test.go]: I followed what was defined in https://github.com/coreos/scripts/blob/master/build_library/disk_layout.json#L29

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,32 @@
+FROM golang:_GOVERSION_
```
Comment (dghubble) [inline Dockerfile]: Can you move this under a `build` directory please. Since its for a build/test image.

File: Jenkinsfile [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+            stage("Build & Test $GOVERSION") {
+                sh '''#!/bin/bash -ex
+
+sudo chmod +x docker_build
+sudo chmod +x test
+sudo chmod +x build
+
+sed -i "s/_GOVERSION_/$GOVERSION/g" Dockerfile
+docker build --rm --tag=test .
+docker run --rm -e TARGET=${GOARCH} -e GOARCH=${GOARCH} -e CGO_ENABLED=${CGO_ENABLED} --privileged -u "$(id -u):$(id -g)" -v /etc/passwd:/etc/passwd:ro -v /etc/group:/etc/group:ro -v "$PWD":/usr/src/myapp -w /usr/src/myapp test sudo -E "PATH=$PATH:/go/bin:/usr/local/go/bin" ./docker_build
```
Comment (dghubble) [inline Jenkinsfile]: There really shouldn't be a need for this docker in docker approach these days. Jenkins can be told to run a particular docker image or Kubernetes pod. I haven't been following this thread, but this all seems needlessly complicated.

File: Jenkinsfile [hunk 14]
```diff
@@ -0,0 +1,69 @@
+#!groovy
```
Comment (dghubble) [inline Jenkinsfile]: Have you considered writing this with declarative pipelines? It does a good job at structuring what goes in a Jenkinsfile so people don't go wild adding tons of test logic here.

File: build [hunk 15]
```diff
@@ -19,5 +19,16 @@ export GOBIN=${PWD}/bin/${GOARCH}
 export GOPATH=${PWD}/gopath
 export CGO_ENABLED=1
 
+if [ "${ACTION-BUILD}" == "NOBUILD" ]; then
```
Comment (dghubble) [inline build]: build script with a NOBUILD option smells.

The actual problem you're trying to solve is that there are now test binaries. This is where using a Makefile becomes a better choice than using this script.

File: build [hunk 16]
```diff
@@ -19,5 +19,16 @@ export GOBIN=${PWD}/bin/${GOARCH}
...
+	exit
+fi
+
 echo "Building ${NAME}..."
 go build -ldflags "${GLDFLAGS}" -o ${GOBIN}/${NAME} ${REPO_PATH}/internal
+
+for D in tests/*; do
+	if [ -d "${D}" ]; then
+		echo "Building ${D}"
+		go build -o ${GOBIN}/${D} ${REPO_PATH}/${D}
```
Comment (dghubble) [inline build]: go install, not build

File: docker_build [hunk 17]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
```
Comment (dghubble) [inline docker_build]: place in `build` directory

File: test [hunk 18]
```diff
@@ -28,7 +28,16 @@ fi
...
 
-echo "Running tests..."
-go test -timeout 60s -cover $@ ${PKG} --race
+if [ "${ACTION:-TEST}" != "COMPILE" ]; then
+    echo "Running tests..."
+    go test -timeout 60s -cover $@ ${PKG} --race
+else
+    PKG=$(cd gopath/src/${REPO_PATH}; go list ./... | \
+    	grep --invert-match vendor)
+    echo "Compiling tests..."
```
Comment (dghubble) [inline test]: Running go tests, compiling is misleading, implying you're producing the test binaries

Comment (arithx) [inline test]: It's running `go test -c` which compiles test binaries to `<package>.test` but doesn't run them.

File: Jenkinsfile [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+    node("${ARCH} && docker") {
+        def CGO = (ARCH == 'arm64') ? 1 : 0
+        withEnv(["TARGET=${ARCH}", "CGO_ENABLED=${CGO}",
+                 "GOARCH=${ARCH}", "GOVERSION=${GOVERSION}"]) {
+            stage("SCM $GOVERSION") {
+                checkout scm
+            }
+
+            stage("Build & Test $GOVERSION") {
+                sh '''#!/bin/bash -ex
```
Comment (dghubble) [inline Jenkinsfile]: Also, can you move this script into the build directory? I want this to be more akin to Jenkinsfile simply calling `make integration-tests` or `./integration-tests` (plausibly usable in other environments), instead of baked into the Jenkinsfile.

File: Dockerfile [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+		e2fsprogs \
+		dosfstools \
+		file \
+	&& rm -rf /var/lib/apt/lists/*
+
+ENV GOLANG_VERSION _GOVERSION_
+
+RUN echo '%jenkins ALL=NOPASSWD:ALL' >> /etc/sudoers;
+
+ENV GOPATH /go
```
Comment (dghubble) [inline Dockerfile]: The golang images should all be setup with the right environment variables I believe. Here's my builder (admittedly different packages are needed):

```
FROM golang:1.8.3-alpine3.6

RUN apk add --no-cache ca-certificates curl bash git make openssh \
  && rm -rf /var/cache/apk/*
```

File: Jenkinsfile (hunk 13 above)
Comment (dghubble) [inline Jenkinsfile]: Stephen mentioned offline, its not for docker in docker, the privileged flag can probably be cleaned up.

File: Jenkinsfile [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+sudo chmod +x docker_build
+sudo chmod +x test
+sudo chmod +x build
+
+sed -i "s/_GOVERSION_/$GOVERSION/g" Dockerfile
```
Comment (dghubble) [inline Jenkinsfile]: Most of our other projects that use this build image strategy maintain the build image Dockerfile in the repo and publish it separately. That way anyone (human or Jenkins alike) can use it to work or debug in the same environment. It also avoids the time or scripting of building it every time.

File: Jenkinsfile (hunk 14 above)
Comment (arithx) [inline Jenkinsfile]: I have considered declarative pipelines. My understanding is that when you are doing things like dynamically creating stages (in this case matrixing between goversions & archs) scripted  is preferred to declarative.

Comment (dghubble) [inline Jenkinsfile]: There is no decision tree, declarative pipelines can include script blocks which container scripted groovy. It just discourages writing too much in scripts. This is not a blocker btw.

File: test (hunk 18 above)
Comment (dghubble) [inline test]: So it is. I'd put that in `build` then and write those test binaries to bin

File: Jenkinsfile (hunk 19 above)
Comment (arithx) [inline Jenkinsfile]: I added the integration-tests script for the later shell block which takes care of actually running the tests.

**Comment by arithx** [comment]
Marking WIP while I'm adding additional test cases and negative tests.

File: tests/validator.go [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+		}
+		sgdiskInfo, err := exec.Command(
+			"sgdisk", "-i", strconv.Itoa(e.Number),
+			imageFile).CombinedOutput()
+		if err != nil {
+			t.Error("sgdisk -i", strconv.Itoa(e.Number), err)
+			return
+		}
+
+		re, err := regexp.Compile("Partition unique GUID: (?P<partition_guid>[\\d\\w-]+)")
```
Comment (bgilbert) [inline tests/validator.go]: `regexp.MustCompile` would be appropriate here.

The repeated code could move into a helper function.

File: tests/validator.go [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+			} else {
+				t.Error("Stat type assertion failed, this will only work on Linux")
+				return
+			}
+		} else {
+			t.Error("Stat type assertion failed, this will only work on Linux")
+			return
+		}
+	} else {
+		if linkInfo.Mode()&os.ModeSymlink == 0 {
```
Comment (bgilbert) [inline tests/validator.go]: `linkInfo.Mode()&os.ModeType != os.ModeSymlink`

File: tests/validator.go [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+			return
+		}
+	}
+	validateNode(t, linkInfo, link.Node)
+}
+
+func validateMode(t *testing.T, path string, mode string) {
+	if mode != "" {
+		sout, err := exec.Command(
+			"stat", "-c", "%a", path).CombinedOutput()
```
Comment (bgilbert) [inline tests/validator.go]: Why not use `os.Stat()`?

File: tests/filesystem.go [hunk 25]
```diff
@@ -0,0 +1,517 @@
...
+func removeFile(t *testing.T, imageFile string) {
+	err := os.Remove(imageFile)
+	if err != nil {
+		t.Log(err)
+	}
+}
+
+func removeMountFolders(t *testing.T, partitions []*types.Partition) {
+	for _, p := range partitions {
+		err := os.RemoveAll(p.MountPath)
```
Comment (bgilbert) [inline tests/filesystem.go]: If I understand correctly, the mountpoints should be empty, so it'd be safer to just make this a `Remove`.

File: tests/filesystem.go [hunk 26]
```diff
@@ -0,0 +1,517 @@
...
+	match := re.FindSubmatch(kpartxOut)
+	if len(match) < 2 {
+		t.Log(string(kpartxOut))
+		t.Fatal("couldn't find device")
+	}
+	return fmt.Sprintf("/dev/mapper/%sp%d", string(match[1]), number)
+}
+
+func writeIgnitionConfig(t *testing.T, config string) string {
+	tmpDir, err := ioutil.TempDir(os.Getenv("TMPDIR"), "config")
```
Comment (bgilbert) [inline tests/filesystem.go]: Just use an empty first argument.

File: tests/filesystem.go [hunk 27]
```diff
@@ -0,0 +1,517 @@
...
+	}
+
+	createPartitionTable(t, imageFile, partitions)
+
+	for counter, partition := range partitions {
+		if partition.TypeCode == "blank" || partition.FilesystemType == "" {
+			continue
+		}
+
+		mntPath, err := ioutil.TempDir(os.Getenv("TMPDIR"), fmt.Sprintf("hd1p%d", counter))
```
Comment (bgilbert) [inline tests/filesystem.go]: First argument blank.

File: tests/filesystem.go [hunk 28]
```diff
@@ -0,0 +1,517 @@
...
+	_ = mustRun(t, "kpartx", args...)
+}
+func formatPartition(t *testing.T, partition *types.Partition) {
+	var mkfs string
+	var opts, label, uuid []string
+
+	switch partition.FilesystemType {
+	case "vfat":
+		mkfs = "mkfs.vfat"
+		if partition.FilesystemLabel != "" {
```
Comment (bgilbert) [inline tests/filesystem.go]: For brevity the conditionals can move out of the individual cases to the `append` lines below.

File: tests/filesystem.go [hunk 29]
```diff
@@ -0,0 +1,517 @@
...
+// kpartxAdd will use kpartx to add partition mappings for all of the partitions
+// contained in the imageFile. This creates devices such as /dev/mapper/loop1p1
+// corresponding to partitions in the imageFile. The loop device name (e.g.
+// "loop1") will be returned.
+func kpartxAdd(t *testing.T, imageFile string) string {
+	args := []string{"-avs", imageFile}
+	_ = mustRun(t, "kpartx", args...)
+	args = []string{"-l", imageFile}
+	kpartxOut := mustRun(t, "kpartx", args...)
+	re, err := regexp.Compile("/dev/(?P<device>[\\w\\d]+)")
```
Comment (bgilbert) [inline tests/filesystem.go]: Just use `MustCompile`.

File: tests/filesystem.go [hunk 30]
```diff
@@ -0,0 +1,517 @@
...
+		}
+		args := []string{partition.Device, partition.MountPath}
+		_ = mustRun(t, "mount", args...)
+	}
+}
+
+func updateTypeGUID(t *testing.T, partition *types.Partition) {
+	switch partition.TypeCode {
+	case "coreos-resize":
+		partition.TypeGUID = "3884DD41-8582-4404-B9A8-E9B84F2DF50E"
```
Comment (bgilbert) [inline tests/filesystem.go]: These could be converted to a map lookup.

File: tests/types/types.go [hunk 31]
```diff
@@ -0,0 +1,205 @@
...
+
+type Link struct {
+	Node
+	Target string
+	Hard   bool
+}
+
+type Node struct {
+	Name  string
+	Path  string
```
Comment (bgilbert) [inline tests/types/types.go]: Consider combining `Name` and `Path` for ergonomics, and possibly adding `Name` and `Directory` methods for convenience.  At the very least, `Path` should be renamed to `Dir` or `Directory`, since a file path is always inclusive of its name.

File: tests/validator.go [hunk 32]
```diff
@@ -0,0 +1,252 @@
...
+func validateMode(t *testing.T, path string, mode string) {
+	if mode != "" {
+		fileInfo, err := os.Stat(path)
+		if err != nil {
+			t.Error("Error running stat on node", path, err)
+			return
+		}
+
+		// only taking last 4 digits for comparison
+		actualMode := strconv.FormatInt(int64(fileInfo.Mode()), 8)
```
Comment (bgilbert) [inline tests/validator.go]: The canonical way to do this is `& 07777`.

File: tests/filesystem.go [hunk 33]
```diff
@@ -0,0 +1,517 @@
...
+		if partition.FilesystemLabel != "" {
+			label = []string{"-n", partition.FilesystemLabel}
+		}
+		if partition.FilesystemUUID != "" {
+			uuid = []string{"-i", partition.FilesystemUUID}
+		}
+	case "ext2", "ext4":
+		mkfs = "mke2fs"
+		opts = []string{
+			"-t", partition.FilesystemType, "-b", "4096",
```
Comment (bgilbert) [inline tests/filesystem.go]: As previously mentioned, this and the tune2fs command need a comment saying that they're mirroring `disk_util`.

File: tests/positive/storage/reformat_filesystem.go [hunk 34]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/coreos/ignition/tests/register"
+	"github.com/coreos/ignition/tests/types"
+)
+
+func init() {
+	register.Register(register.PositiveTest, ReformatToBTRFS())
+	register.Register(register.PositiveTest, ReformatToXFS())
+}
+
+func ReformatToBTRFS() types.Test {
```
Comment (bgilbert) [inline tests/positive/storage/reformat_filesystem.go]: From previous review iteration:

> Why just these two FS types, why only 2.0 configs, and why not also test setting the UUID?

File: .gitignore [hunk 35]
```diff
@@ -1,3 +1,6 @@
 gopath/
 bin/
 *.swp
+*.img
```
Comment (arithx) [inline .gitignore]: Yes they are, the image files are too large to fit in the tmpfs. Do you have a suggestion for working around that?

File: tests/blackbox_test.go [hunk 36]
```diff
@@ -0,0 +1,218 @@
...
+func outer(t *testing.T, test types.Test, negativeTests bool) {
+	PreCleanup(t)
+	t.Log(test.Name)
+
+	tmpDirectory, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatalf("failed to create a temp dir: %v", err)
+	}
+	os.Chmod(tmpDirectory, 0755)
+	_ = os.Setenv("TMPDIR", tmpDirectory)
```
Comment (arithx) [inline tests/blackbox_test.go]: Whenever we reference any files/directories inside of `TMPDIR` we do a `filepath.Join` `os.Getenv`. If this were to fail the tests should still run just pointing in a different location.

Should we fail for this if it's not a breaking failure?

File: tests/filesystem.go (hunk 27 above)
Comment (arithx) [inline tests/filesystem.go]: Currently we're creating one `TMPDIR` per test and then all the mount points and configs are placed as subdirectories inside of this. Should this not be the case?

File: tests/blackbox_test.go [hunk 37]
```diff
@@ -0,0 +1,205 @@
...
+			updateTypeGUID(t, part)
+		}
+		setOffsets(disk.Partitions)
+		for _, part := range test.Out[i].Partitions {
+			updateTypeGUID(t, part)
+		}
+		setOffsets(test.Out[i].Partitions)
+
+		// NOTE: the image file is written into cwd because sgdisk fails when
+		// the file is located in /tmp/
```
Comment (arithx) [inline tests/blackbox_test.go]: Moved `TMPDIR` to `/var/tmp`, removed this comment, and updated `disk.ImageFile` to be inside of `TMPDIR`.

File: tests/filesystem.go (hunk 26 above)
Comment (arithx) [inline tests/filesystem.go]: See comment relating to `mntPath` on same topic.

File: .gitignore (hunk 35 above)
Comment (bgilbert) [inline .gitignore]: Yeah, put your scratch directory in `/var/tmp`.  You may want to do it only if TMPDIR isn't custom, so the user can still override.

File: tests/filesystem.go (hunk 27 above)
Comment (bgilbert) [inline tests/filesystem.go]: Leaving the first argument blank should fall back to `os.TempDir` which will check TMPDIR for you.

File: tests/blackbox_test.go (hunk 36 above)
Comment (bgilbert) [inline tests/blackbox_test.go]: Setenv should really never fail, but if it does, better to fail and make someone fix the problem than to cause surprising behavior silently.
